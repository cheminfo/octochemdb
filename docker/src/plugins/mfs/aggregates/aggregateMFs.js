import debugLibrary from '../../../utils/Debug.js';
/**
 * @description Aggregate in function of unique mfs in compounds collection
 * @param {*} connection mongo connection
 * @returns {Promise} returns mfs collection
 */
export async function aggregate(connection) {
  const debug = debugLibrary('aggregateMFs');
  // get compounds collection and progress
  const collection = await connection.getCollection('compounds');
  const progressCompounds = await connection.getProgress('compounds');
  // set progress to aggregating
  const progress = await connection.getProgress('mfs');
  await connection.setProgress(progress);
  try {
    if (progressCompounds.seq === progress.seq) {
      debug('Aggregation up-to-date');
      return;
    }
    progress.state = 'aggregating';
    await connection.setProgress(progress);
    debug(`Need to aggregate ${await collection.count()} entries`);
    let result = await collection.aggregate(
      [
        {
          $match: {
            'data.nbFragments': 1,
            'data.charge': 0,
            'data.mf': {
              $regex: '^[^]]+$',
            },
          },
        }, // one fragment, no charge and no isotopes
        {
          $project: {
            _id: 0,
            mf: '$data.mf',
            em: '$data.em',
            unsaturation: '$data.unsaturation',
            atoms: '$data.atoms',
          },
        },
        {
          $group: {
            _id: '$mf',
            em: { $first: '$em' },
            unsaturation: { $first: '$unsaturation' },
            atoms: { $first: '$atoms' },
            count: { $sum: 1 },
          },
        },
        { $out: 'mfs_tmp' },
      ],
      {
        allowDiskUse: true,
        maxTimeMS: 60 * 60 * 1000 * 12, // 6h
      },
    );
    await result.hasNext();
    const temporaryCollection = await connection.getCollection('mfs_tmp');
    // rename temporary collection to mfs
    await temporaryCollection.rename('mfs', { dropTarget: true });
    // set progress to aggregated
    progress.dateEnd = new Date();
    progress.seq = progressCompounds.seq;
    progress.state = 'aggregated';
    await connection.setProgress(progress);

    return result;
  } catch (e) {
    progress.state = 'error';
    await connection.setProgress(progress);
    if (connection) {
      debug(e.message, { collection: 'mfs', connection, stack: e.stack });
    }
  }
}
