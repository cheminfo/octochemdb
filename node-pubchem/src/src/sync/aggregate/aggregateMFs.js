import Debug from 'debug';

import {
  COMPOUNDS_COLLECTION,
  MFS_COLLECTION,
} from '../../util/PubChemConnection.js';

const debug = Debug('aggregateMFs');

export default async function aggregateMFs(connection) {
  const collection = await connection.getCollection(COMPOUNDS_COLLECTION);
  debug('Need to aggregate', await collection.countDocuments(), 'entries');
  let result = await collection.aggregate(
    [
      { $match: { nbFragments: 1, charge: 0, mf: /^[^\]]+$/ } }, // one fragment, no charge and no isotopes
      { $project: { _id: 0, mf: 1, em: 1, unsaturation: 1, atom: 1 } },
      {
        $group: {
          _id: '$mf',
          em: { $first: '$em' },
          unsaturation: { $first: '$unsaturation' },
          atom: { $first: '$atom' },
          total: { $sum: 1 },
        },
      },
      { $out: 'mfs' },
    ],
    {
      allowDiskUse: true,
      maxTimeMS: 60 * 60 * 1000, // 1h
    },
  );
  await result.hasNext();

  let mfsCollection = await connection.getCollection(MFS_COLLECTION);
  await mfsCollection.createIndex({ em: 1 });
  await mfsCollection.createIndex({ total: 1 });

  return result;
}
