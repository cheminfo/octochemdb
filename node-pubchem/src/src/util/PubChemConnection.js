'use strict';

const delay = require('delay');
const { MongoClient } = require('mongodb');
const debug = require('debug')('PubChemConnection');

function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function getProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.updateOne({ _id: progress._id }, { $set: progress });
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    debug('Starting new database construction.');
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  } else {
    if (progress.state === 'update') {
      debug('First importation has been completed. Should only update.');
    } else {
      debug(`Continuing first importation from ${progress.seq}.`);
    }
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};

module.exports = PubChemConnection;
