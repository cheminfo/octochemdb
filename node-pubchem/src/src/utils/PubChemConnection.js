import Debug from 'debug';
import delay from 'delay';
import { MongoClient } from 'mongodb';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
