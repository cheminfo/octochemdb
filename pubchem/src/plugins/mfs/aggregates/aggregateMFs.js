import Debug from '../../../utils/Debug.js';

const debug = Debug('aggregateMFs');

export async function aggregate(connection) {
  const collection = await connection.getCollection('compounds');

  const progressCompounds = await connection.getProgress('compounds');
  const progress = await connection.getProgress('mfs');
  if (progressCompounds.seq === progress.seq) {
    debug('Aggregation up-to-date');
    return;
  }

  debug(`'Need to aggregate' ${await collection.countDocuments()} 'entries'`);
  let result = await collection.aggregate(
    [
      {
        $match: {
          'data.nbFragments': 1,
          'data.charge': 0,
          'data.mf': /^[^\]]+$/,
        },
      }, // one fragment, no charge and no isotopes
      {
        $project: {
          _id: 0,
          mf: '$data.mf',
          em: '$data.em',
          unsaturation: '$data.unsaturation',
          atom: '$data.atom',
        },
      },
      {
        $group: {
          _id: '$mf',
          em: { $first: '$em' },
          unsaturation: { $first: '$unsaturation' },
          atom: { $first: '$atom' },
          total: { $sum: 1 },
        },
      },
      { $out: 'mfs' },
    ],
    {
      allowDiskUse: true,
      maxTimeMS: 60 * 60 * 1000, // 1h
    },
  );
  await result.hasNext();

  let mfsCollection = await connection.getCollection('mfs');
  await mfsCollection.createIndex({ em: 1 });
  await mfsCollection.createIndex({ total: 1 });

  progress.seq = progressCompounds.seq;
  await connection.setProgress(progress);

  return result;
}
