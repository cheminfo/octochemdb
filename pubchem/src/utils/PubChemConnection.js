import delay from 'delay';
import { MongoClient } from 'mongodb';

import Debug from './Debug.js';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getLogs = async function getCollection(
  collectionName,
) {
  const logsCollection = await this.getLogsCollection();
  const _id = `${collectionName}`;
  let logs = await logsCollection.find({ _id }).next();
  if (logs === null) {
    let sources = [];
    logs = {
      _id,
      sources,
      sourcesHash: md5(JSON.stringify(sources)),
      dataStart: 0,
      dataEnd: 0,
      startSequenceID: 123,
      endSeqneceID: 123,
      status: 'updating',
    };
    await logsCollection.insertOne(logs);
  }
  return logs;
};
PubChemConnection.prototype.setLogs = async function setLogs(logs) {
  const collection = await this.getLogsCollection();
  await collection.replaceOne({ _id: logs._id }, logs);
};
PubChemConnection.prototype.getLogsCollection = async function getCollection() {
  return (await this.getDatabase()).collection('logs');
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
