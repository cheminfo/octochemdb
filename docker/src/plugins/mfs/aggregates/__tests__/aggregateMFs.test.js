import { PubChemConnection } from '../../../../utils/PubChemConnection.js';
import { aggregate } from '../aggregateMFs';

jest.setTimeout(10000);
test('Aggregation mfs', async () => {
  const connection = new PubChemConnection();
  const collections = await connection.getCollectionNames();
  while (collections.includes('compounds') === false) {
    delay(1000);
  }
  await aggregate(connection);
  const collection = await connection.getCollection('mfs');
  const collectionEntry = await collection
    .find({
      _id: 'C12H9N3S2',
    })
    .limit(1);
  const result = await collectionEntry.next();
  expect(result).toMatchSnapshot();
  if (connection) {
    await connection.close();
  }
});
