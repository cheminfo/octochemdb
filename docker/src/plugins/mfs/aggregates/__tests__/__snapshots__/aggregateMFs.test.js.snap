// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Aggregation mfs 1`] = `
{
  "_id": "C12H9N3S2",
  "data": {
    "atoms": {
      "C": 12,
      "H": 9,
      "N": 3,
      "S": 2,
    },
    "count": 1,
    "em": 259.02378965215996,
    "unsaturation": 10,
  },
}
`;
